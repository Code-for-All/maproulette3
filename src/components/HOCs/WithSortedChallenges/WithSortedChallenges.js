import React, { Component } from 'react';
import PropTypes from 'prop-types'
import _get from 'lodash/get'
import _findIndex from 'lodash/findIndex'
import _sortBy from 'lodash/sortBy'
import _reverse from 'lodash/reverse'
import _isEmpty from 'lodash/isEmpty'
import _omit from 'lodash/omit'
import WithChallengeSearch from '../WithSearch/WithChallengeSearch'
import { SORT_NAME, SORT_CREATED, ALL_SORT_OPTIONS }
       from '../../../services/Search/Search'

const FEATURED_POINTS = -1
const SAVED_POINTS = -2

export const sortChallenges = function(props, challengesProp='challenges') {
  const sortCriteria = _get(props, 'searchSort.sortBy')

  let sortedChallenges = props[challengesProp]

  if (sortCriteria === SORT_NAME) {
    sortedChallenges = _sortBy(sortedChallenges, 'name')
  }
  else if (sortCriteria === SORT_CREATED) {
    sortedChallenges = _reverse(_sortBy(sortedChallenges, 'created'))
  }
  else {
    // default "smart" sort. Prioritizes featured and user-saved challenges
    const savedChallenges = _get(props, 'user.savedChallenges', [])

    sortedChallenges = _sortBy(sortedChallenges, challenge => {
      let score = 0
      score += challenge.featured ? FEATURED_POINTS : 0
      score += _findIndex(savedChallenges, {id: challenge.id}) !== -1 ?
               SAVED_POINTS : 0
      return score
    })
  }

  return sortedChallenges
}


export default function(WrappedComponent,
                        challengesProp='challenges',
                        outputProp) {
  class WithSortedChallenges extends Component {

    render() {
      const sortedChallenges = sortChallenges(this.props, challengesProp)

      if (_isEmpty(outputProp)) {
        outputProp = challengesProp
      }

      return <WrappedComponent sortOptions = {ALL_SORT_OPTIONS}
                               {...{[outputProp]: sortedChallenges}}
                               {..._omit(this.props, outputProp)} />
    }
  }

  WithSortedChallenges.propTypes = {
    user: PropTypes.object,
    challenges: PropTypes.array,
  }

  return WithChallengeSearch(WithSortedChallenges)
}
