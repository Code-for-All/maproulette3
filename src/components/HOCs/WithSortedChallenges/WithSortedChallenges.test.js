import React, { Component } from 'react'
import _findIndex from 'lodash/findIndex'
import WithSortedChallenges from './WithSortedChallenges'

let basicProps = null
let WrappedComponent = null
let normalChallenge = null
let featuredChallenge = null
let savedChallenge = null
let featuredAndSavedChallenge = null

beforeEach(() => {
  normalChallenge = {id: 1}
  featuredChallenge = {id: 2, featured: true}
  savedChallenge = {id: 3 }
  featuredAndSavedChallenge = {id: 4, featured: true}

  basicProps = {
    user: {
      savedChallenges: []
    },
    challenges: [
      normalChallenge,
      featuredChallenge,
      savedChallenge,
      featuredAndSavedChallenge,
    ]
  }

  WrappedComponent = WithSortedChallenges(
    () => <div className="child" />
  )
})

test("all challenges are passed through to the wrapped component", () => {
  const wrapper = shallow(
    <WrappedComponent {...basicProps} />
  )

  expect(wrapper.props().challenges.length).toEqual(basicProps.challenges.length)

  expect(wrapper).toMatchSnapshot()
})

test("featured challenges come ahead of normal challenges", () => {
  const wrapper = shallow(
    <WrappedComponent {...basicProps} />
  )

  const sortedChallenges = wrapper.props().challenges

  expect(
    sortedChallenges.indexOf(featuredChallenge)
  ).toBeLessThan(
    sortedChallenges.indexOf(normalChallenge)
  )

  expect(wrapper).toMatchSnapshot()
})

test("challenges saved by the user come ahead of normal challenges", () => {
  basicProps.user.savedChallenges = [savedChallenge]

  const wrapper = shallow(
    <WrappedComponent {...basicProps} />
  )

  const sortedChallenges = wrapper.props().challenges

  expect(
    sortedChallenges.indexOf(savedChallenge)
  ).toBeLessThan(
    sortedChallenges.indexOf(normalChallenge)
  )

  expect(wrapper).toMatchSnapshot()
})

test("saved and featured challenges come ahead of everything else", () => {
  basicProps.user.savedChallenges = [savedChallenge, featuredAndSavedChallenge]

  const wrapper = shallow(
    <WrappedComponent {...basicProps} />
  )

  const sortedChallenges = wrapper.props().challenges

  expect(
    sortedChallenges.indexOf(featuredAndSavedChallenge)
  ).toBeLessThan(
    sortedChallenges.indexOf(featuredChallenge)
  )

  expect(
    sortedChallenges.indexOf(featuredAndSavedChallenge)
  ).toBeLessThan(
    sortedChallenges.indexOf(savedChallenge)
  )

  expect(
    sortedChallenges.indexOf(featuredAndSavedChallenge)
  ).toBeLessThan(
    sortedChallenges.indexOf(normalChallenge)
  )

  expect(wrapper).toMatchSnapshot()
})
