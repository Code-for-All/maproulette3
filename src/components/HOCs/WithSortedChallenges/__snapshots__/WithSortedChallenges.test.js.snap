// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`all challenges are passed through to the wrapped component 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WithSortedChallenges
    challenges={
        Array [
            Object {
              "id": 1,
            },
            Object {
              "featured": true,
              "id": 2,
            },
            Object {
              "id": 3,
            },
            Object {
              "featured": true,
              "id": 4,
            },
          ]
    }
    user={
        Object {
            "savedChallenges": Array [],
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "challenges": Array [
        Object {
          "featured": true,
          "id": 2,
        },
        Object {
          "featured": true,
          "id": 4,
        },
        Object {
          "id": 1,
        },
        Object {
          "id": 3,
        },
      ],
      "user": Object {
        "savedChallenges": Array [],
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "challenges": Array [
          Object {
            "featured": true,
            "id": 2,
          },
          Object {
            "featured": true,
            "id": 4,
          },
          Object {
            "id": 1,
          },
          Object {
            "id": 3,
          },
        ],
        "user": Object {
          "savedChallenges": Array [],
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`challenges saved by the user come ahead of normal challenges 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WithSortedChallenges
    challenges={
        Array [
            Object {
              "id": 1,
            },
            Object {
              "featured": true,
              "id": 2,
            },
            Object {
              "id": 3,
            },
            Object {
              "featured": true,
              "id": 4,
            },
          ]
    }
    user={
        Object {
            "savedChallenges": Array [
              Object {
                "id": 3,
              },
            ],
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "challenges": Array [
        Object {
          "id": 3,
        },
        Object {
          "featured": true,
          "id": 2,
        },
        Object {
          "featured": true,
          "id": 4,
        },
        Object {
          "id": 1,
        },
      ],
      "user": Object {
        "savedChallenges": Array [
          Object {
            "id": 3,
          },
        ],
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "challenges": Array [
          Object {
            "id": 3,
          },
          Object {
            "featured": true,
            "id": 2,
          },
          Object {
            "featured": true,
            "id": 4,
          },
          Object {
            "id": 1,
          },
        ],
        "user": Object {
          "savedChallenges": Array [
            Object {
              "id": 3,
            },
          ],
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`featured challenges come ahead of normal challenges 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WithSortedChallenges
    challenges={
        Array [
            Object {
              "id": 1,
            },
            Object {
              "featured": true,
              "id": 2,
            },
            Object {
              "id": 3,
            },
            Object {
              "featured": true,
              "id": 4,
            },
          ]
    }
    user={
        Object {
            "savedChallenges": Array [],
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "challenges": Array [
        Object {
          "featured": true,
          "id": 2,
        },
        Object {
          "featured": true,
          "id": 4,
        },
        Object {
          "id": 1,
        },
        Object {
          "id": 3,
        },
      ],
      "user": Object {
        "savedChallenges": Array [],
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "challenges": Array [
          Object {
            "featured": true,
            "id": 2,
          },
          Object {
            "featured": true,
            "id": 4,
          },
          Object {
            "id": 1,
          },
          Object {
            "id": 3,
          },
        ],
        "user": Object {
          "savedChallenges": Array [],
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`saved and featured challenges come ahead of everything else 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WithSortedChallenges
    challenges={
        Array [
            Object {
              "id": 1,
            },
            Object {
              "featured": true,
              "id": 2,
            },
            Object {
              "id": 3,
            },
            Object {
              "featured": true,
              "id": 4,
            },
          ]
    }
    user={
        Object {
            "savedChallenges": Array [
              Object {
                "id": 3,
              },
              Object {
                "featured": true,
                "id": 4,
              },
            ],
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "challenges": Array [
        Object {
          "featured": true,
          "id": 4,
        },
        Object {
          "id": 3,
        },
        Object {
          "featured": true,
          "id": 2,
        },
        Object {
          "id": 1,
        },
      ],
      "user": Object {
        "savedChallenges": Array [
          Object {
            "id": 3,
          },
          Object {
            "featured": true,
            "id": 4,
          },
        ],
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "challenges": Array [
          Object {
            "featured": true,
            "id": 4,
          },
          Object {
            "id": 3,
          },
          Object {
            "featured": true,
            "id": 2,
          },
          Object {
            "id": 1,
          },
        ],
        "user": Object {
          "savedChallenges": Array [
            Object {
              "id": 3,
            },
            Object {
              "featured": true,
              "id": 4,
            },
          ],
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
